internal/async/gen.go

```diff
@@ -111,15 +107,20 @@ func NewUnbounded{{.Name}}Chan() *Unbounded{{.Name}}Chan {
 		// queue is garbage collected.
 		q := make([]{{.Type}}, 0, 1<<10)
 		for {
-			e, ok := <-ch.in
-			if !ok {
-				close(ch.out)
-				return
+			select {
+			case e, ok := <-ch.in:
+				if !ok {
+					close(ch.out)
+					return
+				}
+				q = append(q, e)
+			case <-ch.close:
+				goto closed
```

Comment:
Can this be done without `goto`? This makes code much harder to read, even `break label` is difficult. 
In this instance it is code to execute before return, so instead it could be defined as a helper function and called before a return I think?

Reply:
Done.